### ISSUE TITLE
[user]: Elyssa Collins

### ISSUE THREAD

**Issue by elyssac02** [c-h-david/rapid#27]
### Your Name

Elyssa Collins

### Your Institution

Center for Geospatial Analytics, North Carolina State University

### Your Location

Raleigh, North Carolina, United States

### Your Use Case

I'm using RAPID to develop climate projections (from 32 global climate models) of streamflow throughout the NHDPlus network across a moderate a high greenhouse gas scenario. These streamflow data will then be used with a terrain-based inundation model to investigate how climate change will influence the spatial distribution of future flood probabilities. The study domain is still being decided, but probably across a few HUC6 watersheds in North Carolina. The temporal domain is 2006 - 2100 (2006 - 2025 for the reference period and 2026 - 2100 for the projected period). The runoff data used for RAPID optimization comes from [Livneh et al., 2015](https://www.nature.com/articles/sdata201542) and the runoff data used for the climate change part of the study comes from [Vano et al., 2020](https://gdo-dcp.ucllnl.org/downscaled_cmip_projections/techmemo/LOCA_BCSD_hydrology_tech_memo.pdf) --  specifically, we are using the [LOcalized Constructed Analogs CMIP5 hydrology projections](https://gdo-dcp.ucllnl.org/downscaled_cmip_projections/dcpInterface.html). 

### Why RAPID?

We chose RAPID because it's open source, it can be applied to large spatial domains/is computationally efficient, has strong documentation, and has many papers to look to for optimization and applications of the model. The fact that RAPID uses mapped rivers as its computational elements was also necessary for us. While RAPID being dockerized wasn't initially something we considered, it has proven to be very beneficial. 

### How do you use RAPID?

We use your Docker images. I use RAPID on my local machine (Windows 10, but I also use Ubuntu 20.04.1 with WSL) and a high performance computer. For running RAPID on HPC, I converted the Docker image to a Singularity image. 

### Are you aware of our philosphy for open source?

- [X] I have read the last paragraph of Section 6.3 in [David et al., 2016](https://doi.org/10.1002/2015EA000142) (the one starting with
 "Finally, and contrary to common belief, open source software does
not mean...")

**Comment by c-h-david** [comment]
@elyssac02 
Hi Elyssa, thank you so much for taking the time to fill out the form!  It's great to hear about your exciting project. I had heard of Singularity images but have yet to look into it. Do you perhaps have an example somewhere? Maybe something similar to the [RAPID Dockerfile](https://github.com/c-h-david/rapid/blob/master/Dockerfile)? Does Singularity have a hub similar to [Docker Hub for RAPID](https://hub.docker.com/r/chdavid/rapid/tags)?  We could perhaps consider adding a Singularity capability to the RAPID repository. Thanks again!

**Comment by elyssac02** [comment]
@c-h-david 
Of course! Sorry for taking a while to get back to you on this. A little while ago, [I created a repo](https://github.com/elyssac02/singularity-on-hpc-henry2) for how to use Singularity on HPC (with some specific instructions for my university's cluster). In the repo, you'll see that it's relatively straightforward to convert a Docker image to a Singularity image and this is what I did with RAPID. However, I did have to slightly modify a couple lines of code (e.g., in `rapid_main.F90`), rebuild RAPID, and save a copy of the image as a .tar file ([see here](https://github.com/elyssac02/singularity-on-hpc-henry2/blob/main/README.md#modifying-contents-in-the-image)) to get it to work with running multiple optimization procedures and simulations simultaneously on HPC. Since the `rapid_namelist` is read directly, running multiple procedures/simulations simultaneously also required writing to local scratch on the compute node (although, I'm sure there are other approaches to this). 

I don't have an example on GitHub yet, however, I would be happy to either create one in [my Singularity repo](https://github.com/elyssac02/singularity-on-hpc-henry2) or I could create a pull request containing example scripts here in the RAPID repo (might make more sense). I have examples for running 1) a single optimization procedure, 2) a single regular run, 3) simultaneous optimization procedures, and 4) simultaneous regular runs. The simultaneous procedures/runs involves simultaneously deploying multiple Singularity containers with different inputs as tasks using a tool called [pynodelauncher](https://github.com/ncsu-landscape-dynamics/pynodelauncher) along with writing to local scratch, as mentioned above.

I think the equivalent to a Dockerfile is a [Singularity Recipe](https://sylabs.io/guides/2.6/user-guide/container_recipes.html), however, I didn't have to create this since I convert the Docker image to a Singularity image. It looks like there once was a Singularity Hub, [but looks like they decided to go read-only in 2021](https://singularityhub.github.io/singularityhub-docs/2021/going-read-only/). In any case, I think including instructions in the RAPID repository for converting the Docker image to a Singularity image could be helpful. If I create a pull request, I could also include the .tar file (same contents as the current RAPID docker image, just with a couple of slight modifications to a couple of scripts) used to build the Singularity image on HPC. This .tar file should work for any use case, I would just need to specify some specific instructions for paths. 

Hope this answers your question and is helpful! Let me know if you would like me to create a pull request and if so, what you would like me to include (e.g., the .tar file, example scripts + documentation, etc.).
